### PR TITLE
!!! FEATURE: Translatable forms

### PR DESCRIPTION
**Pull request by kdambekalns**
This change allows for forms to be translated using XLIFF as elsewhere
in Flow. Form element labels, placeholders, option labels and more can
be translated.

Instructions are included in a new documentation chapter.

This is marked as breaking: if you changed the "translationPackage"
earlier to customize validation error messages, you will need to use the
new "validationErrorTranslationPackage" setting for that now. A code
migration to adjust that is included, run:

    ./flow flow:core:migrate  --version 20160601101500 --package-key ...

### CODE REVIEW

**Comment by kdambekalns** [comment]
I have to admit I did not run any of the form tests.

And this changes the public api in a way: if someone changed the `translationPackage` earlier to customise validation error messages, this will no longer work, since there is a new setting now (´validationErrorTranslationPackage`). Ideas for how to deal with this (better) are welcome!

**Comment by aertmann** [comment]
nice to see this sorted out, thanks! a configuration migration could be provided.. and it should be mentioned in the PR description

**Comment by kitsunet** [comment]
Nice, but I guess a configuration migration is definitely needed as well as a breaking mark and possibly next major then?

**Comment by kdambekalns** [comment]
Added migration, marked breaking, adjusted PR message. No problem to release a new major, IMHO.

**Comment by aertmann** [comment]
👍 by reading

**Comment by kdambekalns** [comment]
maybe @bwaidelich could have a look?

**Comment by bwaidelich** [comment]
FYI: I started looking into this but it seems to break the TYPO3.FormBuilder package.. Looking into it right now

File: Configuration/Settings.yaml [hunk 1]
```diff
@@ -7,6 +7,10 @@ TYPO3:
     presets:
       default:
         title: 'Default'
+        renderingOptions:
+          # Package with XLIFF files for translation of form labels, placeholders, ...
+          # If you need custom messages, copy Resources/Private/Translations/en/Main.xlf to your package and adjust this setting
+          translationPackage: 'TYPO3.Form'
```
Comment (bwaidelich) [inline Configuration/Settings.yaml]: This setting is not expected/evaluated here. And the now missing `formElementTypes.'TYPO3.Form:Base'.renderingOptions.translationPackage` setting leads to an exception ```#1416832309: The current package key can't be resolved.```.
So instead, we should keep this setting (line 31) and *add* the `validationErrorTranslationPackage` setting on the same level

File: Documentation/translating-forms.rst [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+.. code-block:: yaml
+
+    TYPO3:
+      Form:
+        presets:
+          default:
+            formElementTypes:
+              'TYPO3.Form:Base':
+                renderingOptions:
+                  translationPackage: 'AcmeCom.SomePackage'
```
Comment (bwaidelich) [inline Documentation/translating-forms.rst]: here it's correct ;)

File: Migrations/Code/Version20160601101500.php [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+     * Recurse into the given preset and rename translationPackage to validationErrorTranslationPackage
+     *
+     * @param array $preset
+     * @return array
+     */
+    public function renameTranslationPackage(array &$preset)
+    {
+        foreach ($preset as $key => $value) {
+            if (is_array($value)) {
+                if (array_key_exists('translationPackage', $value)) {
```
Comment (bwaidelich) [inline Migrations/Code/Version20160601101500.php]: here it's wrong again, and the migration should not replace the existing `translationPackage`-Setting but rather add the `validationErrorTranslationPackage`, no?

File: Resources/Private/Form/Page.html [hunk 4]
```diff
@@ -1,7 +1,7 @@
 {namespace form=TYPO3\Form\ViewHelpers}
 <fieldset>
 	<f:if condition="{page.label}">
-		<legend>{page.label}</legend>
+		<legend>{page.label -> f:translate(id: 'forms.pages.{page.identifier}.label', package: '{element.renderingOptions.translationPackage}')}</legend>
```
Comment (bwaidelich) [inline Resources/Private/Form/Page.html]: it must be `{page.renderingOptions.translationPackage}` here, not `{element...`

**Comment by bwaidelich** [comment]
I added a few comments inline. Apart from those, that change leads to weird behaviors when omitting placeholders:

image

**Comment by bwaidelich** [comment]
BTW: Regarding

> that change leads to weird behaviors when omitting placeholders

The technical reason for this is this nasty line here:

https://github.com/neos/flow-development-collection/blob/master/TYPO3.Fluid/Classes/TYPO3/Fluid/ViewHelpers/TranslateViewHelper.php#L119

-> by default the `placeholder` property is an empty string which evaluates to FALSE there.. muh

File: Migrations/Code/Version20160601101500.php (hunk 3 above)
Comment (kdambekalns) [inline Migrations/Code/Version20160601101500.php]: No, it should change an existing setting. Until now, that was only used for error message translation, so…

Comment (bwaidelich) [inline Migrations/Code/Version20160601101500.php]: But if the setting is not there you'll get the mentioned exception

**Comment by kdambekalns** [comment]
@bwaidelich Thanks for the review, should all be fixed.

**Comment by bwaidelich** [comment]
@kdambekalns not there yet.. I'll push a follow-up

File: Migrations/Code/Version20160601101500.php (hunk 3 above)
Comment (kdambekalns) [inline Migrations/Code/Version20160601101500.php]: But it exists in the Form package itself, so… no? I removed it and my site worked as expected (not showing translations). Or does that happen when not using a preset based on `default`?

**Comment by bwaidelich** [comment]
FYI: I added a  custom `TranslateElementProperty`-ViewHelper as discussed.. To work properly, this requires https://github.com/neos/flow-development-collection/pull/377 to be merged

**Comment by bwaidelich** [comment]
*ping* anyone?

**Comment by kdambekalns** [comment]
Have it open in my browser since Wednesday … :)

**Comment by bwaidelich** [comment]
> Have it open in my browser since Wednesday

Haha. No stress ;)

**Comment by bwaidelich** [comment]
FYI: #391 has been merged in the meantime

**Comment by kdambekalns** [comment]
Very nice, thanks @bwaidelich for the collaboration!

**Comment by bwaidelich** [comment]
whoop whoop, thanks to *you* for initiating this long awaited feature!

**Pull request merged**